src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp

```diff
@@ -0,0 +1,934 @@
+#include <Processors/QueryPlan/ITransformingStep.h>
+#include <Processors/QueryPlan/AggregatingStep.h>
+#include <Processors/QueryPlan/ExpressionStep.h>
+#include <Processors/QueryPlan/JoinStep.h>
+#include <Processors/QueryPlan/ArrayJoinStep.h>
+#include <Processors/QueryPlan/CreatingSetsStep.h>
+#include <Processors/QueryPlan/CubeStep.h>
+#include <Processors/QueryPlan/ReadFromMergeTree.h>
+#include <Processors/QueryPlan/SortingStep.h>
+#include <Processors/QueryPlan/TotalsHavingStep.h>
+#include <Processors/QueryPlan/DistinctStep.h>
+#include <Processors/QueryPlan/UnionStep.h>
+#include <Processors/QueryPlan/WindowStep.h>
+#include <Interpreters/ActionsDAG.h>
+#include <Interpreters/ArrayJoinAction.h>
+#include <Interpreters/InterpreterSelectQuery.h>
+#include <Interpreters/TableJoin.h>
+#include <Common/typeid_cast.h>
+#include <Storages/StorageMerge.h>
+#include <Functions/IFunction.h>
+#include <DataTypes/DataTypeAggregateFunction.h>
+#include <Columns/IColumn.h>
+#include <stack>
+
+
+namespace DB::QueryPlanOptimizations
+{
+
+ISourceStep * checkSupportedReadingStep(IQueryPlanStep * step)
+{
+    if (auto * reading = typeid_cast<ReadFromMergeTree *>(step))
+    {
+        /// Already read-in-order, skip.
+        if (reading->getQueryInfo().input_order_info)
+            return nullptr;
+
+        const auto & sorting_key = reading->getStorageMetadata()->getSortingKey();
+        if (sorting_key.column_names.empty())
+            return nullptr;
+
+        return reading;
+    }
+
+    if (auto * merge = typeid_cast<ReadFromMerge *>(step))
+    {
+        const auto & tables = merge->getSelectedTables();
+        if (tables.empty())
+            return nullptr;
+
+        for (const auto & table : tables)
+        {
+            auto storage = std::get<StoragePtr>(table);
+            const auto & sorting_key = storage->getInMemoryMetadataPtr()->getSortingKey();
+            if (sorting_key.column_names.empty())
+                return nullptr;
+        }
+
+        return merge;
+    }
+
+    return nullptr;
+}
+
+QueryPlan::Node * findReadingStep(QueryPlan::Node & node)
+{
+    IQueryPlanStep * step = node.step.get();
+    if (auto * reading = checkSupportedReadingStep(step))
+        return &node;
+
+    if (node.children.size() != 1)
+        return nullptr;
+
+    if (typeid_cast<ExpressionStep *>(step) || typeid_cast<FilterStep *>(step) || typeid_cast<ArrayJoinStep *>(step))
+        return findReadingStep(*node.children.front());
+
+    if (auto * distinct = typeid_cast<DistinctStep *>(step); distinct && distinct->isPreliminary())
+        return findReadingStep(*node.children.front());
+
+    return nullptr;
+}
+
+/// FixedColumns are columns which values become constants after filtering.
+/// In a query "SELECT x, y, z FROM table WHERE x = 1 AND y = 'a' ORDER BY x, y, z"
+/// Fixed columns are 'x' and 'y'.
+using FixedColumns = std::unordered_set<const ActionsDAG::Node *>;
```

Comment:
Maybe better to rename it into FixedNodes, because as I undestand nodes here are not only column, but can be function.